### DISCUSSION TITLE
Explain different types of inheritance in C#

### DISCUSSION THREAD

**Discussion by mishelshaji** [mishelshaji/Learn-DotNet#13] [Q&A] [answered]
Explain what is inheritance, benefits of inheritance, Real-life example of inheritance, explain with code snippets.

**Comment by Jiby-Joseph-123** [comment, marked as answer]
# INHERITANCE

Taking the properties of one class to another class is inheritance. Reusability of code is the main advantage of inheritance
The inheritance concept is based on a base class and a derived class. 

Base class - is the class from which features are to be inherited into another class.
Derived class - it is the class in which the base class features are inherited.

# Types of inheritance

- Single Inheritance
- Hierarchical Inheritance
- Multi-level Inheritance
- Multiple Inheritance

# Single Inheritance

Child class(derived class) derived from a single parent class(base class)
```csharp
// base class
class Animal
{
    public string Name;
    public string Color;
     public Animal(string name, string color)
    {
        this.Name = name;
        this.Color = color;
    }
    public void display()
    {
        Console.WriteLine("Iam an animal");
    }
}
// derived class

class Dog : Animal
{
    public int Age;
    public Dog(string name, string color,int age):base(name,color)
    {
    this.Age = age;
    }
    public void display()
    {
        Console.WriteLine($"Iam  {Name}, my color is {Color} age {Age}");
    }
}

class main
{
    public static void Main()
    {
        var obj = new Animal("Puppy", "Black");
        obj.display();
        var obj1 = new Dog("Pupp", "Brown",2);
        obj1.display();
    }
}
```

# Multilevel Inheritance

When one class inherits another class it is further inherited by another class. One of the derived classes is a parent of another class.
```csharp
// base class
class Animal
{
    public string Name;
    public string Color;
    public Animal(string name, string color)
    {
        this.Name = name;
        this.Color = color;
    }
    public void display()
    {
        Console.WriteLine("Iam an animal");
    }
}
// derived class
class Dog : Animal
{
    public int Age;
    public Dog(string name, string color, int age) : base(name, color)
    {
        this.Age = age;
    }
    public void display()
    {
        Console.WriteLine($"Iam  {Name}, my color is {Color} age {Age}");
    }
}
class BabyDog : Dog
{
    public string NickName;
    public BabyDog(string name, string color, int age,string nickName) : base(name,color,age)
    {
        NickName = nickName;
    }

    public void display()
    {
        Console.WriteLine($"Iam  {Name}, my color is {Color} age {Age} my nick name is {NickName}");
    }
}

class main
{
    public static void Main()
    {
        var obj = new Animal("Puppy", "Black");
        obj.display();
        var obj1 = new Dog("Pupp", "Brown", 2);
        obj1.display();
        var obj3 = new BabyDog("Pupp", "Brown", 2, "Buno");
        obj3.display();
    }
}
```
# Hierarchical Inheritance

More than one class is derived from a base class.
```csharp
// base class
class Animal
{
    public string Name;
    public string Color;
    public Animal(string name, string color)
    {
        this.Name = name;
        this.Color = color;
    }
    public void display()
    {
        Console.WriteLine("Iam an animal");
    }
}
// derived class
class Dog : Animal
{
    public int Age;
    public Dog(string name, string color, int age) : base(name, color)
    {
        this.Age = age;
    }
    public void display()
    {
        Console.WriteLine($"Iam  {Name}, my color is {Color} age {Age}");
    }
}
// parent class of Tiger class is Animal
class Tiger : Animal
{
    public Tiger(string name, string color) : base(name, color) 
    {

    }
    public void display()
    {
        Console.WriteLine($"Iam  {Name}, my color is {Color}  And I am a wild animal.");
    }
}
class main
{
    public static void Main()
    {
        var obj = new Animal("Puppy", "Black");
        obj.display();
        var obj1 = new Dog("Pupp", "Brown", 2);
        obj1.display();
        var obj3 = new Tiger("Pupp", "Brown");
        obj3.display();
    }
}
```
# Multiple Inheritance

A class derived from more than one parent class. C# doesn't support multiple inheritances. Inheritance is used for multiple inheritances.
```csharp
// base class
using System.Drawing;
using System.Xml.Linq;

interface Animal
{
    public void displayAnimal();
  
}
// parent class
interface Domestic
{
    public void displayDomestic();
}
    
    // derived class
    class Dog : Animal , Domestic
{
    public void displayAnimal()
    {
            Console.WriteLine("It is a animal");
    }
        public void displayDomestic()
        {
            Console.WriteLine("It is a  domestic animal");
        }
    }
class main
{
    public static void Main()
    {
        var obj = new Dog();
        obj.displayAnimal();
        obj.displayDomestic();
        
    }
}
```

**Reply by mishelshaji** [reply]
Please note the following points:

- The class name should use PascalCasing. 
- Remove unused references to `using System.Drawing;` and `using System.Xml.Linq;` namespaces.
- Comments should start with a capital letter and end with a period`.`.
- Code blocks should be properly indented.
- A blank line should be added in between method and class declarations.

**Comment by rashid234** [comment]
Inheritance:-
Inheritance allows us to define a class in terms of another class, which makes it easier to create and maintain an application.This also provides an opportunity to reuse the code functionality and speeds up implementation time. Accessing properties of one class to other class. There should be more than one class is required to establish inheritance one is base class and other is derived class
base class :- is there class where features are taken to other class(parent class)
derived class :- is there class that take the features of other class. (child class)

Types of inheritance:-
single inheritance : - child class derived from only one parent
multiple inheritance :- child class contains more than one parent class(does not support in c#)
multilevel inheritance :- child class derived from a parent class and that class is again derived from other parent class
Hierarchical Inheritance :- more than one class is derived from a parent class

**Reply by mishelshaji** [reply]
Please try to give an example when you answer a question.

**Comment by ARAVIND0407** [comment]
Acquiring the properties of one class into another class is called inheritance
In C# there are 
- Single
    It is the type of inheritance in which there is one base class and one derived class.

- Multilevel
    When one class is derived from another derived class then this type of inheritance is called multilevel inheritance.

- Hierarchical
    This is the type of inheritance in which there are multiple classes derived from one base class.

**Reply by mishelshaji** [reply]
Please try to give an example when you answer a question.

**Comment by sarath534** [comment]
# Inheritance in C
 It is a process of acquiring the properties of one class to another class. Inheritance provide reusability of code by allowing us to extend the class.

## Types of Inheritance:
###### Note:
Base class - is the class from which features are to be inherited into another class.
Derived class - it is the class in which the base class features are inherited.

### Single Inheritance :

- In this types of inheritance we have only one base class and one derived class.
```csharp
public class A //base class    
{  
    public string Name()  
    {  
        return "Name of class A";  
    }  
}  
public class B : A //derived class    
{  
    public string Register()  
    {  
        Name();                       ////derived class method    
        return "Register of class A";  
    }  
}
```
### Hierarchical inheritance
- In this type of inheritance we have multiple classes derived from one base class. This type of inheritance is used when there is a need for one class properties in many different class.
```csharp
class A  //base class    
{  
    public string msg()  
    {  
        return "this is A class Method";  
    }  
}  
class B : A  
{  
    public string info()  
    {  
        msg();  
        return "this is B class Method";  
    }  
    class C : A  
    {  
        public string getinfo()  
        {  
            msg();  
            return "this is B class Method";  
        }  
    }  
} 
```
### Multilevel inheritance:
- When one class is derived from another derived class then it is called as multilevel inheritance.
```csharp
public class Student  
{  
    public string StudentID()  
    {  
        return "this is the Student class";  
    }  
}  
public class Subject :  Student
{  
    public string SubjectID()  
    {  
        StudentID();  
        return "this is the Subject Class";  
    }  
}  
public class Marks : Subject 
{  
    public string MarkScored()  
    {  
        StudentID();  
        SubjectID();  
        return "this is the Mark Class";  
    }  
} 
```
### Multiple inheritance using Interfaces:
- C# does not support multiple inheritances of classes. To overcome this problem we can use interfaces.
- Multiple inheritance is the process by which a class is derived from two base class.
```csharp
public interface IA //ineterface  1    
{  
    string setImgs(string a);  
}  
public interface IB  //Interface 2    
{  
    int getAmount(int Amt);  
}  
public class ICar : IA, IB //implementatin    
{  
    public int getAmount(int Amt)  
    {  
        return 100;  
    }  
    public string setImgs(string a)  
    {  
        return "this is the car";  
    }  
} 
```

**Reply by mishelshaji** [reply]
Please follow coding conventions.

- A blank line should be added between each method.
- Method name should be in pascal casing.
- Never write comments on the same line after writing a statement.
- A comment should start with `//`.
- Never write string values directly in the return statement.

**Comment by Nandana-04** [comment]
# Inheritance

Inheritance is one of the fundamental attributes of object-oriented programming . It is a process in which one object acquires all the properties and behaviors of its parent object automatically.

- The class whose members are inherited is called Base class. 
- The class which inherits the members of another class is called Derived class .
- The derived class is the specialized class for the base class.

Benefits of Inheritance :-

- Inheritance promotes reusability. When a class inherits or derives another class, it can access all the functionality of inherited class .
- As the existing code is reused, it leads to less development and maintenance costs .
- Inheritance makes the sub classes follow a standard interface .
- Inheritance helps to reduce code redundancy and supports code extensibility .

Types of Inheritance :-

- Single Inheritance
- Hierarchical Inheritance
- Multilevel Inheritance
- Multiple Inheritance
- Hybrid Inheritance

### Single Inheritance
It is the type of inheritance in which there is one base class and one derived class .
```csharp
class Fruit
    {
    public string Name = "Apple";
    }
class fruits : Fruit
    {
        public string place = "Jammu Kashmir";
    }
fruits choice = new fruits();
Console.WriteLine(choice.Name + " : " + choice.place);
```
### Hierarchical Inheritance
It is the type of inheritance in which derived classes inherit from a single base class .
```csharp

 class Fruit
    {
        public string Name;
        public string PlaceName;
    }

class Apple : Fruit
    {
        public int numOfApples ;
    }
    class Orange : Fruit
    {
        public int numOfOranges = 4;
    }

Apple myApple = new Apple();
myApple.Name = "Green Apple";
myApple.PlaceName = "Jammu Kashmir";
myApple.numOfApples = 5;

Console.WriteLine(myApple.Name + " is from " + myApple.PlaceName + " and we have " + myApple.numOfApples);

Orange myOrange = new Orange();
myOrange.Name = "Orange";
myOrange.PlaceName = "Southeastern France";
myOrange.numOfOranges = 4;

Console.WriteLine(myOrange.Name + " is from " + myOrange.PlaceName + " and we have " + myOrange.numOfOranges);
```
### Multilevel Inheritance
It is the type of inheritance in which a derived class inherits from a base class, which in turn inherits from another base class .
```csharp
 class Parent
    {
        public string name = "Ram";
    }
class Child : Parent
    {
        public string Name1 = "Rahul";
    }
class GrandChild : Child 
    {
        public string Name2 = "Riyoo";
    }
GrandChild children = new GrandChild();
Console.WriteLine("GrandFather : {0} , Son : {1} , GrandSon : {2}", children.name ,children.Name1 , children.Name2);
```
### Multiple Inheritance
It is the type of inheritance in which a class can inherit from multiple base classes .
This is not directly supported in C#, but can be achieved using interfaces .
```csharp
 interface Cat
    {
        void displayCat();
    }

    interface Dog
    {
        void displayDog();
    }
class Animal : Cat , Dog
    {
        public void displayCat()
        {
            Console.WriteLine("Cat..");
        }

        public void displayDog()
        {
            Console.WriteLine("Doggg..");
        }
    }

Animal animal = new Animal();
animal.displayCat();
animal.displayDog();
```
### Hybrid Inheritance
 A combination of more than one type of inheritance .

**Comment by jeweljose22** [comment]
Inheritance is a mechanism in object-oriented programming (OOP) that allows a class to inherit properties and methods from another class. This allows for code reuse and a more organized and efficient codebase.

Benefits of inheritance include:

Code reuse: Classes that inherit from other classes can use the properties and methods of those classes without having to rewrite them.
Method overriding: Child classes can override the methods of their parent class to provide their own implementation.

Method overloading: Child classes can overload the methods of their parent class to provide their own implementation with different number or type of parameters.

Simplifying code: Inheritance can make code more organized and easier to understand.

A real-life example of inheritance is a car class that inherits from a vehicle class. The car class would have properties and methods specific to cars (e.g. number of doors, car make and model) and would inherit properties and methods from the vehicle class (e.g. number of wheels, speed).
```csharp
class Vehicle
{
    public int NumberOfWheels { get; set; }
    public void Move() { /* code to move the vehicle */ }
}

class Car : Vehicle
{
    public int NumberOfDoors { get; set; }
    public void HonkHorn() { /* code to honk the horn */ }
}
```

**Comment by SonuSanilK** [comment]
**Inheritance**

Acquiring the properties of one class into another class is called inheritance. Inheritance provides reusability by allowing us to extend an existing class.
Inheritance are of different types. They are:
- Single Inheritance
It is the type of inheritance in which there is one base class and one derived class.

class A(Base class)
{
}
class B:A(derived class)
{
}

-Hierarchical inheritance

This is the type of inheritance in which there are multiple classes derived from one base class. This type of inheritance is used when there is a requirement of one class feature that is needed in multiple classes.

Class A(Base class)
{
}
class B:A
{
}
class C:A
{
}
class D:A
{
}

-Multilevel inheritance

C# does not support multiple inheritances of classes. To overcome this problem we can use interfaces.

interface A
{
}
interface B
{
}
class C:A,B
{
}

**Comment by amalapaul** [comment]
## INHERITANCE

Acquiring the properties of one class into another class is called inheritance. Inheritance provides reusability by allowing us to extend an existing class.
Inheritance is a programming concept of one or more child classes receiving fields methods from a common parent like how people inherit genes from their parents .
Base class is the class from which the members are inherited and the class that inherits those members is called the derived class or child class.

An example is shown below,

```csharp
  // Base class.
  class Animal { 

    public string name;

    public void display() {
      Console.WriteLine("I am an animal");
    }
    
  } 
  
  // Derived class of Animal 
  class Goat : Animal {
    
    public void getName() {
      Console.WriteLine("My name is " + name);
    }
  }

  class Program {

    static void Main(string[] args) {

      // Object of derived class.
      Goat Jamnapyari = new Goat();

      //Method of base class.
      Jamnapyari.name = "AmmuKutty";
      Jamnapyari.display();


      Jamnapyari.getName();

      Console.ReadLine();
    }

  }
}
```

### Advantages of Inheritance:

- Inheritance promotes reusability. When a class inherits or derives another class, it can access all  functionalities of inherited class.

- Reusability enhanced reliability. The base class code will be already tested and debugged.

-  Less development and maintenance costs.

- Inheritance helps to reduce code redundancy


### The following are the types of inheritance in C#.

-Single Inheritance
-Multi Level Inheritance
-Hierarchical Inheritance
-Multiple Inheritance

### Single inheritance 
 
It is the type of inheritance in which there is one base class and one derived class.

```csharp
public class TrainArrivalInfo //base class    
{  
    public string Arrival()  
    {  
        return "Train is arrived.";  
    }  
}  
public class DepartureInfo : TrainArrivalInfo //Derived class. 
{  
    public string Departure()  
    {  
        Arrival();                       //Derived class method.   
        return "Train is departured.";  
    }  
}  
```

### Hierarchical inheritance
 
This is the type of inheritance in which there are multiple classes derived from one base class. 

```csharp
class A  //Base class    
{  
    public string msg()  
    {  
        return "this is A class Method";  
    }  
}  
class B : A  
{  
    public string info()  
    {  
        msg();  
        return "this is B class Method";  
    }  
    class C : A  
    {  
        public string getinfo()  
        {  
            msg();  
            return "this is B class Method";  
        }  
    }  
}   
 ```

### Multilevel inheritance
 
When one class is derived from another derived class then this type of inheritance is called multilevel inheritance.


```csharp
public class Flower  
{  
    public string flowerdet()  
    {  
        return "These are flowers,";  
    }  
}  
public class Fruits : Flower 
{  
    public string fruitdet()  
    {  
        flowerdet();  
        return "These are fruits.";  
    }  
}  
public class Fish : Fruit 
{  
    public string fishdet()  
    {  
        flowerdet();  
        fruitdet();  
        return "these are fishes";  
    }  
}   
```

### Multiple inheritance using Interfaces
 
C# does not support multiple inheritances of classes so we use interfaces.

```csharp
public interface IA  //Interface A  
{  
    string setImgs(string a);  
}  
public interface IB  //Interface B   
{  
    int getAmount(int Amt);  
}  
public class ICar : IA, IB    
{  
    public int getAmount(int Amt)  
    {  
        return 100;  
    }  
    public string setImgs(string a)  
    {  
        return "This is the car";  
    }  
} 
```

**Comment by akhilamp123** [comment]
**INHERITANCE**

By using inheritance in C#, we can build a new class from an existing one. It is a fundamental component of Object-Oriented Programming (OOP).The base class is the class that a new class is constructed from (parent or superclass). And the new class is referred to as a derived class (child or subclass). Fields and methods from the base class are passed down to the derived class. 

**Types Of Inheritance**

- Single Inheritance
- Multilevel Inheritance
- Hierarchical Inheritance
- Multiple Inheritance
- Hybrid Inheritance

**Single Inheritance**

In single inheritance, a single derived class inherits from a single base class

```csharp
class Parent
    {
        public void DisplayParentsAB()
        {
            Console.WriteLine(" A and B are Parents.");
        }
    }

    class Son : Parent
    {
        public void DisplaySonC()
        {
            Console.WriteLine(" C is Son");
        }
    }

Son s = new Son();
s.DisplaySonC();
s.DisplayParentsAB();
`````

**Multilevel Inheritance**

In multilevel inheritance, a derived class inherits from a base and the same derived class acts as a base class for another class.

```csharp
class GrandParent
    {
        public void DisplayGrandParentsAB()
        {
            Console.WriteLine(" A and B are GrandParents.");
        }
    }

    class Parent : GrandParent
    {
        public void DisplayParentsCD()
        {
            Console.WriteLine(" C and D are parents.");
        }
    }
    class Child : Parent
    {
        public void DisplayChildE()
        {
            Console.WriteLine(" E is child ");
        }
    }
Child c = new Child();
c.DisplayChildE();
c.DisplayParentsCD();
c.DisplayGrandParentsAB();
`````

**Hierarchical Inheritance**

In hierarchical inheritance, multiple derived classes inherit from a single base class

````csharp
class Parent
    {
        public void DisplayParentsAB()
        {
            Console.WriteLine(" A and B are Parents.");
        }
    }

    class ChildC : Parent
    {
        public void DisplayChildC()
        {
            Console.WriteLine(" C is the child of parent A & B.");
        }
    }
    class ChildD : Parent
    {
        public void DisplayChildD()
        {
            Console.WriteLine(" D is child of parent A & B ");
        }
    }
ChildC c = new ChildC();
c.DisplayChildC();

ChildD d = new ChildD();
d.DisplayChildD();

````

**Multiple Inheritance**

In multiple inheritance, a single derived class inherits from multiple base classes.c# doesn't support multiple inheritance. However, we can achieve multiple inheritance through interfaces.

**Hybrid Inheritance**

Hybrid inheritance is a combination of two or more types of inheritance. The combination of multilevel and hierarchical inheritance is an example of hybrid inheritance.

**Advantages of Inheritance**

- Inheritance promotes reusability. When a class inherits or derives another class, it can access all the functionality of inherited class.

- Reusability enhanced reliability. The base class code will be already tested and debugged.
 
- Inheritance makes the sub classes follow a standard interface.

- Inheritance helps to reduce code redundancy and supports code extensibility.
 
- Inheritance facilitates creation of class libraries.

**Comment by sreelakshmiChandran** [comment]
# Inheritance
- Inheritance is the principle of the oop that inherits the properties of one class to another class.
- It allows the reusability of the code.
- Inheritance is based on derived class and base class
- Base class is known as the parent class and the derived class is known as a child where the class derives from the parent class.
 
## Types of inheritance
- Single inheritance
      Single child class for a parent class. Only one class derives from a base class.
- Multi-level inheritance
      When one class is derived from another derived class then it is called multi-level inheritance. One child class is the parent of another child class. 
- Hierarchical inheritance
      Multiple child class from a parent class is called hierarchical inheritance. 
- Multiple inheritances
      A class derived from more than one parent class. C# doesn't support multiple inheritances. Inheritance is used for multiple inheritances.
 
## Single inheritance
```csharp
public class Student //base class    
{  
    public string Name()  
    {  
        return "Student name";  
    }  
}  
public class StudentRegister : Student //derived class    
{  
    public string Register()  
    {  
        Name();                       ////derived class method    
        return "Register of Student ";  
    }  
}
```
## Multi-level inheritance
```csharp
public class Student  
{  
    public string StudentName()  
    {  
        return "Student name";  
    }  
}  
public class Subject :  Student
{  
    public string Subject()  
    {  
        public class();  
        return "Subject ";  
    }  
}  
public class Mark : Subject 
{  
    public string MarkScored()  
    {  
        StudentName();  
        Subject();  
        return "Mark";  
    }  
} 
```
## Hierarchical inheritance
```csharp
class Student  //base class    
{  
    public string Name()  
    {  
        return "Student name";  
    }  
}  
class RegisterNumber : Student  
{  
    public string info()  
    {  
        msg();  
        return "RegisterNumber";  
    }  
    class Subject : Student  
    {  
        public string getinfo()  
        {  
            msg();  
            return "Subject";  
        }  
    }  
} 
```

## Multiple inheritances
- Thus the c# doesn't support multiple inheritances, interfaces are used to overcome this problem.
```csharp
interface Student
{
    public void displayStudent();
  
}
// parent class
interface Subject
{
    public void displaySubject();
}
    
    // derived class
    class Mark : Student , Subject
{
    public void displayStudent()
    {
            Console.WriteLine("Student Name");
    }
        public void displaySubject()
        {
            Console.WriteLine("Subject name");
        }
    }
class main
{
    public static void Main()
    {
        var obj = new Mark();
        obj.displayStudent();
        obj.displaySubject();
        
    }
}
```

**Comment by AshinDamodaran** [comment]
INHERITANCE
Taking the properties of one class to another class is inheritance. Reusability of code is the main advantage of inheritance
The inheritance concept is based on a base class and a derived class.

Base class - is the class from which features are to be inherited into another class.
Derived class - it is the class in which the base class features are inherited.

**Types of inheritance**
- Single Inheritance
- Hierarchical Inheritance
- Multi-level Inheritance
- Multiple Inheritance

**Comment by SreelakshmiSajeev** [comment]
#Inheritance : The process of inheriting the properties of another class by a class is called inheritance. The different types of inheritance in C# are as follows:

-1. Single Inheritance : This type of inheritance contains only one base class and derived class.

#Example.

```C#
using System;

namespace MyApplication
{
   class Animal
  {
    public void Display()
    {
      Console.WriteLine("Hai Iam an animal");
     }
   }
  class Dog : Animal
  {
    public void Print()
    {
      Console.WriteLine("Hi I am a dog...........");
    }
  }

  class Program
  {
    static void Main(string[] args)
    {
      Dog dog = new Dog();  
      dog.print();
  
    }
  }
}

-2.Multilevel Inheritance : Deriving a class a from another class is called multilevel inheritance.

#Example.

```C#
using System;

namespace MyApplication
{
   class Animal
  {
    public void Display()
    {
      Console.WriteLine("Hai Iam an animal");
     }
   }
  class Dog : Animal
  {
    public void Print()
    {
      Console.WriteLine("Hi I am a dog...........");
    }
  }
  
class cat : Dog
  {
    public void Print1()
    {
      Console.WriteLine("Hi I am a cat...........");
    }
  }
  class Program
  {
    static void Main(string[] args)
    {
      Cat cat = new cat();
      cat.Display();
      cat.Print();
      cat.Print1();
  
    }
  }
}

-3.Hierarchical Inheritance : Here the multiple  classes are derived from one base class.

#Example.

```C#
using System;

namespace MyApplication
{
   class Animal
  {
    public void Display()
    {
      Console.WriteLine("Hai Iam an animal");
     }
   }
  class Dog : Animal
  {
    public void Print()
    {
      Console.WriteLine("Hi I am a dog...........");
    }
  }
  
class cat : Animal
  {
    public void Print1()
    {
      Console.WriteLine("Hi I am a cat...........");
    }
  }
  class Program
  {
    static void Main(string[] args)
    {
      Cat cat = new cat();
      cat.Display();
      cat.Print();
      cat.Print1();
  
    }
  }
}

-4.Multiple Inheritance : Here one base class contains more than one derived class. But it is  not supported by the C#. To overcome this it is implemented by using interfaces.

#Example.

 ```C#
using System;

namespace MyApplication
{
   public interface Idog
  {
    string color(string c);
  }
   public interface ICat
  {
    string name(string n);
  }
  
class Animal : Idog,Icat
  {
     public string color(string c)  
    {  
        return "Iam browmn in color";  
    }  
     public string name(string n)  
    {  
        return "Iam Browniee";  
    }  
  class Program
  {
    static void Main(string[] args)
    {
     Animal obj = new Animal();
     obj.color;
     obj.name;
  
    }
  }

**Comment by Samaria200** [comment]
### Inheritance 

Inheritance is one of the fundamental attributes of object-oriented programming. It allows you to define a child class that reuses (inherits), extends, or modifies the behavior of a parent class. 
 The class whose members are inherited is called the **base class**.
 The class that inherits the members of the base class is called the **derived class**.

Types of inheritance includes :

1. Single Inheritance
2. Hierarchical Inheritance
3. Multi level Inheritance
4. Multiple Inheritance 

### Single Inheritance

It is the type of inheritance in which there is one base class and one derived class.
 ```csharp
public class Human 
    {  
       public void eat() { Console.WriteLine("Eating..."); }  
   }  
   public class Child : Human
   {  
       public void cry() { Console.WriteLine("Crying..."); }  
   }  
   class TestInheritance 
{  
       public static void Main(string[] args)  
        {  
            Child kid1 = new Child() ;  
            kid1.eat();  
            kid1.cry();  
        }  
    }  
```
### Multi Level Inheritance 

When one class inherits another class which is further inherited by another class, it is known as multi level inheritance in C#.
```csharp
public class Human  
    {  
       public void eat() { Console.WriteLine("Eating..."); }  
   }  
   public class Child : Human  
   {  
       public void cry() { Console.WriteLine("Crying..."); }  
   }  
   public class BabyChild : Child
   {  
       public void sleep() { Console.WriteLine("Sleeping..."); }  
   }  
   class TestingInheritance 
{  
       public static void Main(string[] args)  
        {  
            BabyChild kid = new BabyChild();  
            kid.eat();  
            kid.cry();  
            kid.sleep();  
        }  
    }  
```
### Hierarchical Inheritance 

This is the type of inheritance in which there are multiple classes derived from one base class.
```csharp
public class Human
    {
        public void learn()
        {
            Console.WriteLine("Learn");
        }
    }
    
    public class Adult : Human
    {
        public void work()
        {
            Console.WriteLine("working");
        }
    }
    
    public class Student : Human
    {
        public void study()
        {
            Console.WriteLine("studying");
        }   
    }
    
    public class Program
    {
        public static void Main(string[] args)
        {
            Adult adult1 = new Adult() ;
            Student student1 = new Student() ;
            
            adult1.Adult();
            adult1.Human() ; // accessing parent class
            
            student1.Student
            student1.Human() ;   // accessing parent class
        }
    }
```
### Multiple Inheritance
Multiple Inheritance is not supported in C# due to ambiguity problem. So we use the concept of interfaces.
```csharp
interface Human
{
    public void live();
  
}
// parent class
interface Child
{
    public void sleep();
}
    
    // derived class
    class Student : Human, Child
{
    public void live()
    {
            Console.WriteLine("It's Human!");
    }
        public void sleep()
        {
            Console.WriteLine("It's a child!");
        }
    }
class main
{
    public static void Main()
    {
        var kid = new Student();
        kid.live() ;
        kid.sleep() ; 
        
    }
}
```

**Comment by Shamla-P** [comment]
# Inheritance
Inheritance is a process in which one object acquires all the properties and behaviors of its parent class. Class which inherits the members of another class is called derived class and the class whose members are inherited is called base class. 
Benefit of inheritance is we can reuse, extend or modify the attributes and behaviors which is defined in other class.
Main advantage is code reusability, we can reuse the members of our parent class. So, there is no need to define the member again. So less code is required in the class.

**Comment by AnitaMaryThomas** [comment]
### **Inheritance**

- Inheritance is a process in which one object acquires all the features and behaviors of its parent object automatically. So we can reuse, extend or modify the attributes and behavoirs which is defined in other class.
- The class whose members are inherited is called derived class.
- The class whose members are inherited is called base class.

**#Types of Inheritance**

- The inheritance mainly classified into four. They are:

1. Single Inheritance
2. Multiple Inheritance
3. Multi-Level Inheritance
4. Hierarchical Inheritance

**1. Single Inheritance**

- When one class inherits another class, it is known as single level inheritance.

**#Example:**
```csharp
using System;
public class Student
{
public int Mark = 50;
}
public class Data: Student
{
public int InternalMark = 19;
}
class SingleInheritance
{
public static void Main(string[] args)
{
Data data = new Data();
Console.WriteLine("Mark: " + data.Mark);
Console.WriteLine("Mark: " + data.InternalMark);
}
}
```
**2. Muliple Inheritance**

Multiple Inheritance in which a class can inherit from multiple base classes.
It is not directly supported in C#, but can be achieved using interfaces.

**#Example:**
```csharp
interface Lion
{
void displayLion();
}
interface Tiger
{
void displayTiger();
}
class Animal : Lion , Tiger
{
public void displayLion()
{
Console.WriteLine("Lion");
}
public void displayTiger()
{
Console.WriteLine("Tiger");
}
}
Animal animal = new Animal();
animal.displayLion();
animal.displayTiger();
```

**3. Multi-Level Inheritance**

- When one class inherits another class which is further inherited by another class, it is known as multi-level inheritance.

**#Example:**
```csharp
using System;
public class Animal
{
public void Eat()
{
Console.WriteLine("Eating");
}
public class Lion : Animal
{
public void Sound()
{
Console.WriteLine("Sound...");
}
public class ChildLion : Lion
{
public Color()
{
Console.WriteLine("Color is brown");
}
class TestInheritance
{
public static void Main(string[] args)
{
ChildLion L1 = new ChildLion();
L1.Eat();
L1.Sound();
L1.Color();
}
}
```
**4. Hierarchical Inheritance**

 - Hierarchical inheritance is the type of inheritance which are multiple classes derived from one base class.

**#Example:**
```csharp
class A
{
public string Data()
{
return "A class method";
}
}
class B : A
{
public string GetData()
{
Data();
return "B class method";
}
class C : A
{
public string InfoData()
{
Data();
return "B class method";
}
}
}
```
